File: src/app/services/crud.service.ts
import { Injectable } from "@angular/core";
import { TauriApiService } from "@app/api/api.api.service";
import { Response } from "@entities/response.model";

export interface CrudParams {
  id?: string;
  data?: unknown;
  filter?: unknown;
}

@Injectable({
  providedIn: "root",
})
export class CrudService {
  constructor(private api: TauriApiService) {}

  async execute<T = unknown>(
    operation: string,
    entity: string,
    params: CrudParams = {}
  ): Promise<T> {
    return this.api.invoke<T>("crud_execute", {
      operation,
      entity,
      id: params.id,
      data: params.data,
      filter: params.filter,
    });
  }

  async get<T = unknown>(entity: string, id: string): Promise<T> {
    return this.execute<T>("get", entity, { id });
  }

  async getAll<T = unknown>(entity: string, filter?: unknown): Promise<T[]> {
    const result = await this.execute<{ [key: string]: T[] }>("get_all", entity, { filter });
    return Object.values(result ?? {})[0] ?? [];
  }

  async create<T = unknown>(entity: string, data: unknown): Promise<T> {
    return this.execute<T>("create", entity, { data });
  }

  async update<T = unknown>(entity: string, id: string, data: unknown): Promise<T> {
    return this.execute<T>("update", entity, { id, data });
  }

  async patch<T = unknown>(entity: string, id: string, data: unknown): Promise<T> {
    return this.execute<T>("patch", entity, { id, data });
  }

  async delete(entity: string, id: string): Promise<void> {
    await this.execute("delete", entity, { id });
  }

  async count(entity: string): Promise<number> {
    return this.execute<number>("count", entity, {});
  }

  async exists(entity: string, id: string): Promise<boolean> {
    return this.execute<boolean>("exists", entity, { id });
  }
}

File: src-tauri/src/services/crud_service.rs
use crate::utils::response::Response;
use nosql_orm::prelude::*;
use serde_json::Value;
use std::sync::Arc;

pub struct CrudService {
  provider: Arc<JsonProvider>,
}

impl CrudService {
  pub fn new(provider: JsonProvider) -> Self {
    Self {
      provider: Arc::new(provider),
    }
  }

  pub async fn execute(
    &self,
    operation: &str,
    entity: &str,
    id: Option<&str>,
    data: Option<Value>,
    _filter: Option<Value>,
  ) -> Result<Response<Value>, String> {
    match operation {
      "get" => {
        let id = id.ok_or("ID required for get")?;
        let result = self
          .provider
          .find_by_id(entity, id)
          .await
          .map_err(|e| e.to_string())?;
        match result {
          Some(data) => Ok(Response::success("Found", data)),
          None => Ok(Response::not_found(entity)),
        }
      }
      "get_all" => {
        let results = self
          .provider
          .find_all(entity)
          .await
          .map_err(|e| e.to_string())?;
        Ok(Response::success("Found", Value::Array(results)))
      }
      "create" | "save" => {
        let data = data.ok_or("Data required for create")?;
        let result = self
          .provider
          .insert(entity, data)
          .await
          .map_err(|e| e.to_string())?;
        Ok(Response::created("Created", result))
      }
      "update" => {
        let id = id.ok_or("ID required for update")?;
        let mut data = data.ok_or("Data required for update")?;
        if let Some(obj) = data.as_object_mut() {
          obj.insert("id".to_string(), Value::String(id.to_string()));
        }
        let result = self
          .provider
          .update(entity, id, data)
          .await
          .map_err(|e| e.to_string())?;
        Ok(Response::updated("Updated", result))
      }
      "patch" => {
        let id = id.ok_or("ID required for patch")?;
        let patch = data.ok_or("Patch data required")?;
        let result = self
          .provider
          .patch(entity, id, patch)
          .await
          .map_err(|e| e.to_string())?;
        Ok(Response::updated("Patched", result))
      }
      "delete" => {
        let id = id.ok_or("ID required for delete")?;
        self
          .provider
          .delete(entity, id)
          .await
          .map_err(|e| e.to_string())?;
        Ok(Response::deleted("Deleted", Value::Null))
      }
      "count" => {
        let count = self
          .provider
          .count(entity, None)
          .await
          .map_err(|e| e.to_string())?;
        Ok(Response::success("Count", Value::Number(count.into())))
      }
      "exists" => {
        let id = id.ok_or("ID required for exists")?;
        let exists = self
          .provider
          .exists(entity, id)
          .await
          .map_err(|e| e.to_string())?;
        Ok(Response::success(
          if exists { "Exists" } else { "Not found" },
          Value::Bool(exists),
        ))
      }
      _ => Err(format!("Unknown operation: {}", operation)),
    }
  }
}
